### PR TITLE
Fix #3852: Desync constructing path with scenery in its way.

### LINKED ISSUE #3852: Place path through object does not remove object on server
**OS:** OS X
**Version:** 0.0.5
**Commit/Build:** 504e5e0

Place path through object only removes object on client side, not server side.
- [ ] Reproducible in RCT2 (vanilla)?
- [x] Multiplayer?

**Screenshots / Video:**
1. Building path on server, watching on server(works fine)
   image
2. Building path on server, watching on client(object not deleted - reconnect solves problem)
   image
3. Building path on client without any permission, watching on client(although fence is removed, since client has no permission so there's no effect to server)
   image
4. Building path on client with permission, watching on client(looks fine, but after reconnect object is still there)
   image
5. Building path on client with permission, watching on server(result is same to 2.)
   image

### LINKED ISSUE #6089: Desync when constructing path with scenery in its way.
Connecting footpath with scenery between seems to cause desyncs, the action will in fact never take place on the server.

- [ ] Reproducible in RCT2 (vanilla)?
- [x] Multiplayer?

**Steps to reproduce:**
1. Load the savegame for host.
2. Join as client and connect the missing footpath, scenery will not be removed on server.

**Save game:**
[desync04.sv6.txt](https://github.com/OpenRCT2/OpenRCT2/files/1189074/desync04.sv6.txt)

### PR DESCRIPTION
**Pull request by ZehM4tt**
When footpath was placed only the local client would remove intersecting walls which causes a desync. In order to properly support this for multiplayer I moved the code into the game command instead of being ran from the window code. This required me to create a new function called footpath_place_remove_intersecting which requires an additional direction parameter used from gFootpathConstructDirection since the server doesn't know about the direction. I made use of the unused ebp parameter for the game command to provide the direction.

### CODE REVIEW

File: src/openrct2/world/footpath.c [hunk 1]
```diff
@@ -620,6 +634,11 @@ money32 footpath_place(sint32 type, sint32 x, sint32 y, sint32 z, sint32 slope,
     return game_do_command(x, (slope << 8) | flags, y, (type << 8) | z, GAME_COMMAND_PLACE_PATH, 0, 0);
 }
 
+money32 footpath_place_remove_intersecting(sint32 type, sint32 x, sint32 y, sint32 z, sint32 slope, sint32 flags, sint32 direction)
+{
+    return game_do_command(x, (slope << 8) | flags, y, (type << 8) | z, GAME_COMMAND_PLACE_PATH, 0, (1 << 8) | direction);
```
Comment (duncanspumpkin) [inline src/openrct2/world/footpath.c]: What is `(1<<8)` define it.

File: src/openrct2/world/footpath.c [hunk 2]
```diff
@@ -493,7 +506,8 @@ void game_command_place_footpath(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32 *
         *edx & 0xFF,
         (*ebx >> 8) & 0xFF,
         *ebx & 0xFF,
-        *edi & 0xFF
+        *edi & 0xFF,
+        *ebp
```
Comment (duncanspumpkin) [inline src/openrct2/world/footpath.c]: Follow the convention of the other fields. The game command function should split up the register into the two `uint8` parameters.

File: src/openrct2/world/footpath.c [hunk 3]
```diff
@@ -333,7 +333,7 @@ static money32 footpath_element_update(sint32 x, sint32 y, rct_map_element *mapE
     return gParkFlags & PARK_FLAGS_NO_MONEY ? 0 : gFootpathPrice;
 }
 
-static money32 footpath_place_real(sint32 type, sint32 x, sint32 y, sint32 z, sint32 slope, sint32 flags, uint8 pathItemType)
+static money32 footpath_place_real(sint32 type, sint32 x, sint32 y, sint32 z, sint32 slope, sint32 flags, uint8 pathItemType, sint32 direction)
```
Comment (duncanspumpkin) [inline src/openrct2/world/footpath.c]: See below use two parameters

File: src/openrct2/world/footpath.c [hunk 4]
```diff
@@ -385,6 +385,19 @@ static money32 footpath_place_real(sint32 type, sint32 x, sint32 y, sint32 z, si
         coord.y = y + 16;
         coord.z = map_element_height(coord.x, coord.y);
         network_set_player_last_action_coord(network_get_player_index(game_command_playerid), coord);
+
+        bool useDirection = !!(direction >> 8);
```
Comment (janisozaur) [inline src/openrct2/world/footpath.c]: Looks like you're trying to access the bit that you set when calling this command? In that case please explicitly mask it with the new mask define and then just shift.

File: src/openrct2/world/footpath.c [hunk 5]
```diff
@@ -493,7 +505,9 @@ void game_command_place_footpath(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32 *
         *edx & 0xFF,
         (*ebx >> 8) & 0xFF,
         *ebx & 0xFF,
-        *edi & 0xFF
+        *edi & 0xFF,
+        (*ebp >> 8) & 0xFF,
```
Comment (janisozaur) [inline src/openrct2/world/footpath.c]: Mask with the flag only and then shift. It will make the intentions here much clearer

**Comment by janisozaur** [comment]
Does this solve https://github.com/OpenRCT2/OpenRCT2/issues/3852 ?

**Comment by ZehM4tt** [comment]
Yes, sorry for the duplicate. I knew it was there I was just unable to find it.

**Pull request merged**